### PR TITLE
Adding a basic snapshot tests.

### PR DESCRIPTION
**Pull request by sduskis**

### CODE REVIEW

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+
+public class TestSnapshots extends AbstractTest {
+
+  static final int count = 10;
```
Comment (igorbernstein2) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: Should count be a constant?

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        ResultScanner scanner = clonedTable.getScanner(new Scan())){
+      for(Result result : scanner) {
+        String row = Bytes.toString(result.getRow());
+        Long expected = values.get(row);
+        Long found = Bytes.toLong(result.getValue(COLUMN_FAMILY, QUALIFIER));
+        Assert.assertEquals("row " + row + " not equal", expected, found);
+        values.remove(row);
+      }
+    }
+    Assert.assertTrue("Extra keys found.", values.isEmpty());
```
Comment (igorbernstein2) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: I think the remaining values are missing rather then extra

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import org.apache.hadoop.hbase.protobuf.generated.HBaseProtos.SnapshotDescription;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+
+public class TestSnapshots extends AbstractTest {
```
Comment (kevinsi4508) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: nit:
create table
list snapshots (no snapshot)
snapshot table
list snapshots
get snapshot
get cloned table (does not exist)
clone snapshot
get cloned table (verify the cloned table)

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  static final int count = 10;
+  final byte[] QUALIFIER = dataHelper.randomData("TestSingleColumnValueFilter");
+
+  private final TableName tableName = IntegrationTests.newTestTableName();
+  private final String snapshotName = tableName.getNameAsString() + "_snp";
+  private final TableName clonedTableName =
+      TableName.valueOf(tableName.getNameAsString() + "_clone");
+
+   @Before
+   public void fillTable() throws IOException {
```
Comment (kevinsi4508) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: delete?

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+   }
+
+  @Test
+  @Category(KnownGap.class)
+  public void testSnapshot() throws IOException {
+    Admin admin = getConnection().getAdmin();
+    admin.createTable(
+      new HTableDescriptor(tableName).addFamily(new HColumnDescriptor(COLUMN_FAMILY)));
+
+    Map<String, Long> values = createTable();
```
Comment (kevinsi4508) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: populateTable()?

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    admin.snapshot(snapshotName, tableName);
+    List<SnapshotDescription> snapshots = admin.listSnapshots(snapshotName);
+    Assert.assertEquals(1, snapshots.size());
```
Comment (kevinsi4508) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: Verify TTL at the minimum.

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java (hunk 3 above)
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: Will do.  should we add delete snapshot?

FYI, HBase doesn't really do a get snapshot.  There's only list, and list with regex filtering.

**Comment by kevinsi4508** [comment]
I think we already have integration test that test all the snapshot operations. The focus here is to make sure the inputs are translated correctly by verifying the responses of snapshot operations.

**Comment by sduskis** [comment]
We need integration tests as well to make sure that HBase and Bigtable work correctly together, especially with the instance ID -> cluster ID translation.  Input validation is important, although we have not done that for other simple APIs.  Given that I'd need to do quite a bit of abstraction for the input validation, I'm not sure if that's the best thing to do at this point.

File: bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java (hunk 6 above)
Comment (kevinsi4508) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java]: that sux :-1:

**Comment by kevinsi4508** [comment]
I am not talking about input validation, more like verifying the result of the what we translated and sent.
e.g. We provide ttl  as an input and make sure the response we get back is expected.

**Comment by sduskis** [comment]
I'm not sure how to validate ttls, since HBase doesn't have that concept.  We should be relying on default ttls from HBase, so there really isn't anything to check, even if we wanted to.

**Comment by kevinsi4508** [comment]
TTL as example is a bad example. Is there anything at all we can verify? Perhaps we should verify at the lower lever.

**Comment by sduskis** [comment]
Isn't "verify at lower level" accomplished by the existing integration tests?

**Comment by kevinsi4508** [comment]
Sorry! Didn't make myself clear. I meant make a service call through HBbase
API, and then call V2 API directly to verify. E.g. createSnapshot HBase
call follow by adminClient getSnapshot.

This is obviously a compatibility issue, we do any much as possible to
verify the result.

**Comment by sduskis** [comment]
I addressed all of the comments, as best as I could.  PTAL.

**Comment by kevinsi4508** [comment]
LGTM

**Pull request merged**